### PR TITLE
Keyboard config

### PR DESCRIPTION
**Pull request by jackpot51**
Will make keymaps editable and provide an interface for an application to change them

### CODE REVIEW

**Comment by jackpot51** [comment]
This is ready for review now. I bundled in hardware keyboard color handling, since it required some other changes in this PR. It's a big beast, for sure. I recommend reading each commit independently to get a better idea of the changes

File: tool/src/main.rs [hunk 1]
```diff
@@ -384,6 +392,33 @@ fn main() {
                     process::exit(1);
                 },
             },
+            "keymap" => {
+                let layer = parse_arg(&mut args, "layer");
+                let output = parse_arg(&mut args, "output");
+                let input = parse_arg(&mut args, "input");
+                match args.next() {
+                    Some(value_str) => match u16::from_str_radix(&value_str, 16) {
```
Comment (crawfxrd) [inline tool/src/main.rs:400]: It will be confusing that `value` is parsed as hex but a `0x` prefix cannot be used.

**Review by crawfxrd — CHANGES_REQUESTED** [review]

File: tool/src/main.rs (hunk 1 above)
Comment (jackpot51) [reply tool/src/main.rs]: 0x prefix trimmed in https://github.com/system76/ec/pull/97/commits/008ee25785e515f08dbd4be04a25a940399c53d4

File: layouts.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
+
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply layouts.sh:4]: Added in https://github.com/system76/ec/pull/97/commits/008ee25785e515f08dbd4be04a25a940399c53d4

File: src/board/system76/common/kbled.c [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#include <board/kbled.h>
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/board/system76/common/kbled.c:3]: Added in https://github.com/system76/ec/pull/97/commits/008ee25785e515f08dbd4be04a25a940399c53d4

File: src/ec/it8587e/include/ec/scratch.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#ifndef _EC_SCRATCH_H
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/ec/it8587e/include/ec/scratch.h:3]: Added in https://github.com/system76/ec/pull/97/commits/008ee25785e515f08dbd4be04a25a940399c53d4

File: src/ec/it5570e/include/ec/scratch.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#ifndef _EC_SCRATCH_H
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/ec/it5570e/include/ec/scratch.h:3]: Added in https://github.com/system76/ec/pull/97/commits/008ee25785e515f08dbd4be04a25a940399c53d4

**Review by crawfxrd — APPROVED** [review]

**Comment by jacobgkau** [comment]
On the addw2, this does appear to be working and I am able to use the configuration tool to remap keys on both layers. However, I've had several instances where Fn-F8, Fn-F9. Fn-F11, and Fn-F12 stopped working, even though the configuration tool showed they were still mapped to their respective functions of `BRIGHTNESS_DOWN`, `BRIGHTNESS_UP`, `AIRPLANE_MODE`, and `SUSPEND`. Shutting off and powering on (resetting the EC and losing any customizations I made) brings those four keys back. I will try to find a way to reproduce this reliably.

**Review by jacobgkau — CHANGES_REQUESTED** [review]
After investigating this further:
- Rebooting (not losing the keyboard layout customizations) allows those four Fn shortcuts to work again-- an EC reset is not required to recover from the issue.
- Suspend/resume is what triggers the shortcuts to stop working. Opening the keyboard configurator is not required to reproduce the issue.
- The issue is present on `904cee4-dirty` (with this EC branch), but is not present on `904cee4` (with https://github.com/system76/ec/commit/c752097753cef9cf4f19ac01aa9887eb0688eb99) or on `904cee4-dirty` with EC https://github.com/system76/ec/commit/c9caca51cdaeb1769ad5e1a875a42346c97fdd93 checked out, which leads me to believe this is a regression in this branch.

**Comment by jackpot51** [comment]
@jacobgkau your suspend/resume issue should be fixed in https://github.com/system76/coreboot/pull/37

**Comment by jacobgkau** [comment]
> your suspend/resume issue should be fixed in [system76/coreboot#37](https://github.com/system76/coreboot/pull/37)

Confirmed the Coreboot PR fixes the suspend/resume issue with this PR. This is now working as expected on the addw2, I am still able to remap keys successfully.

On the bonw14, certain keys in the keyboard configurator display `NONE/NONE` when selected. Those keys are: ```F4, F5, F11, F12, PrtSc/SysRq, Insert/ScrLk, Del, Home, End, `/~, 3/#, 7/&, =/+, Backspace, Q, U, O, 7/Home, 9/PgUp, +, A, G, K, Enter, 4/Left, 5, C, N, Up, 1/End, 3/PgDn, Fn, Alt, Ctrl, Left, Down, Right```

Many other keys show me the wrong item when selected. For example, if I select the `F` key, the default mapping displayed is `Layer 0: C, Layer 1: C` (which is incorrect; the F key works normally and types "f".) Attempting to remap the `F` key actually remaps the `C` key. Would that be an issue in the configuration utility, or does that need to be fixed in firmware?

**Comment by jackpot51** [comment]
The layout editor is hardcoded to use the darp6 layout which works on many other of our laptop keyboards, bit not the bonw14, which has a different electrical layout. So what you see is expected

**Review by jacobgkau — APPROVED** [review]

**Pull request merged**